--- a/mad-0.14.2b/libmad/configure	2001-11-04 06:27:11.000000000 +1100
+++ b/mad-0.14.2b/libmad/config.bak	2017-10-05 13:29:59.248899862 +1100
@@ -1669,7 +1669,6 @@
     case "$optimize" in
 	-O|"-O "*)
 	    optimize="-O"
-	    optimize="$optimize -fforce-mem"
 	    optimize="$optimize -fforce-addr"
 	    : #x optimize="$optimize -finline-functions"
 	    : #- optimize="$optimize -fstrength-reduce"
