--- a/mad-0.14.2b/libmad/configure	2018-02-21 14:44:57.980027034 +1100
+++ b/mad-0.14.2b/libmad/configure	2018-02-21 14:49:57.954337917 +1100
@@ -1669,8 +1669,6 @@
     case "$optimize" in
 	-O|"-O "*)
 	    optimize="-O"
-	    optimize="$optimize -fforce-mem"
-	    optimize="$optimize -fforce-addr"
 	    : #x optimize="$optimize -finline-functions"
 	    : #- optimize="$optimize -fstrength-reduce"
 	    optimize="$optimize -fthread-jumps"
